app/src/main/java/com/eis/phoneringer/MainActivity.java

```diff
@@ -1,14 +1,127 @@
+            @Override
+            public void onClick(View view) {
+                sendCommand();
+            }
+        });
+    }
+
+    /**
+     * Sends a ring command to the phone number given by the user
+     */
+    private void sendCommand() {
+        final Context context = getApplicationContext();
+        //Callback when the command is sent
+        SMSSentListener smsSentListener = new SMSSentListener() {
+            @Override
+            public void onSMSSent(SMSMessage message, SMSMessage.SentState sentState) {
+                Toast.makeText(context, getString(R.string.sent_listener_toast) + phoneNumber.getText().toString(), Toast.LENGTH_SHORT).show();
+            }
+        };
+        //Passing to the AppManager a new RingCommand built through the phone number given by the user
+        try {
+            AppManager.getInstance().sendCommand(context, new RingCommand(new SMSPeer(phoneNumber.getText().toString()), passwordManager.getPassword()), smsSentListener);
+        } catch (InvalidTelephoneNumberException e) {
+            Toast.makeText(context, getString(R.string.invalid_phone_number_toast), Toast.LENGTH_SHORT).show();
+        } catch (InvalidSMSMessageException e) {
+            Toast.makeText(context, getString(R.string.invalid_sms_message_toast), Toast.LENGTH_SHORT).show();
+        }
     }
+
+    /**
+     * @return true if the app has both RECEIVE_SMS and SEND_SMS permissions, false otherwise
+     */
+    public boolean checkPermission() {
+        Context context = getApplicationContext();
+        return ((context.checkSelfPermission(Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED) &&
+                (context.checkSelfPermission(Manifest.permission.SEND_SMS) == PackageManager.PERMISSION_GRANTED));
+    }
+
+    /**
+     * Checks if permissions are granted, if not then requests them to the user
+     */
+    public void requestPermission() {
+        if (!checkPermission())
+            requestPermissions(new String[]{Manifest.permission.SEND_SMS, Manifest.permission.RECEIVE_SMS}, 0);
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+        Context context = getApplicationContext();
+        if (grantResults[0] == PackageManager.PERMISSION_GRANTED && grantResults[1] == PackageManager.PERMISSION_GRANTED) {
+            SMSHandler.getInstance().setup(context);
+            SMSHandler.getInstance().setReceivedListener(new ReceivedMessageListener(context));
+        } else {
+            Toast.makeText(context, getString(R.string.permissions_denied_toast), Toast.LENGTH_SHORT).show();
+            //Let's wait the toast ends
+            Handler handler = new Handler();
+            handler.postDelayed(new Runnable() {
+                public void run() {
+                    requestPermission();
+                }
+            }, WAIT_TIME_PERMISSION);
```

Comment:
To be 100% sure that the time is enough I'd have used `Toast.LENGTH_SHORT + WAIT_TIME_PERMISSION`, where extra time is short

Reply:
`Toast.LENGTH_SHORT` return an integer value equals to zero.
If I put it where u want it won't change anything, `WAIT_TIME_PERMISSION` is enough.